components/server/src/user/env-var-service.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { UserDB } from "@gitpod/gitpod-db/lib";
+import { UserEnvVar, UserEnvVarValue } from "@gitpod/gitpod-protocol";
+import { inject, injectable } from "inversify";
+import { Authorizer } from "../authorization/authorizer";
+import { IAnalyticsWriter } from "@gitpod/gitpod-protocol/lib/analytics";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { v4 as uuidv4 } from "uuid";
+import { Config } from "../config";
+
+@injectable()
+export class EnvVarService {
+    constructor(
+        @inject(Config) private readonly config: Config,
+        @inject(UserDB) private readonly userDB: UserDB,
+        @inject(Authorizer) private readonly auth: Authorizer,
+        @inject(IAnalyticsWriter) private readonly analytics: IAnalyticsWriter,
+    ) {}
+
+    async getAllEnvVars(requestorId: string, userId: string): Promise<UserEnvVarValue[]> {
+        await this.auth.checkPermissionOnUser(requestorId, "read_info", userId);
+        const result: UserEnvVarValue[] = [];
+        for (const value of await this.userDB.getEnvVars(userId)) {
+            result.push({
+                id: value.id,
+                name: value.name,
+                value: value.value,
+                repositoryPattern: value.repositoryPattern,
+            });
+        }
+        return result;
+    }
+
+    async setEnvVar(requestorId: string, userId: string, variable: UserEnvVarValue): Promise<void> {
+        await this.auth.checkPermissionOnUser(requestorId, "write_info", userId);
+        const validationError = UserEnvVar.validate(variable);
+        if (validationError) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, validationError);
+        }
+
+        variable.repositoryPattern = UserEnvVar.normalizeRepoPattern(variable.repositoryPattern);
+        const existingVars = (await this.userDB.getEnvVars(userId)).filter((v) => !v.deleted);
```

Comment:
Out of scope of this PR; but might warrant a TODO for the future: This DB access + checks should be guarded by an transaction.

Reply:
why would you want to wrap this call in a transaction?

Reply:
I want to wrap everything in https://github.com/gitpod-io/gitpod/pull/18503/files#diff-c92102dc489ee1e9088df02532884b52accfceab428fedee173083c14fe8ea41R77-R112, because it feels we are trying to maintain some invariant on the list of env vars, but are not guarding against parallelism.

But totally out of scope here :bow:

Reply:
ah got it. FWIW we just discussed splitting this method into two:
- addUserEnvVar
- updateUserEnvVar